File: fix_schema.js
const fs = require('fs');

let schema = fs.readFileSync('apps/api/prisma/schema.prisma', 'utf8');

// 1. RolePermission
schema = schema.replace(
`model RolePermission {
  id           String     @id @default(uuid())
  roleId       String
  permissionId String
  createdAt    DateTime   @default(now())
  updatedAt    DateTime   @updatedAt
  role         Role       @relation(fields: [roleId], references: [id])
  permission   Permission @relation(fields: [permissionId], references: [id])

  @@unique([roleId, permissionId])
}`,
`model RolePermission {
  id           String     @id @default(uuid())
  tenantId     String
  roleId       String
  permissionId String
  createdAt    DateTime   @default(now())
  updatedAt    DateTime   @updatedAt
  tenant       Tenant     @relation(fields: [tenantId], references: [id])
  role         Role       @relation(fields: [roleId], references: [id], onDelete: Cascade)
  permission   Permission @relation(fields: [permissionId], references: [id], onDelete: Cascade)

  @@unique([roleId, permissionId])
  @@index([tenantId])
}`);

// 2. JournalEntryLine cascade and restrict
schema = schema.replace(
`  journalEntry     JournalEntry    @relation(fields: [journalEntryId], references: [id])`,
`  journalEntry     JournalEntry    @relation(fields: [journalEntryId], references: [id], onDelete: Cascade)`);

schema = schema.replace(
`  contact          Contact?        @relation(fields: [contactId], references: [id])`,
`  contact          Contact?        @relation(fields: [contactId], references: [id], onDelete: Restrict)`);

// 3. Invoice relations
schema = schema.replace(
`  contact        Contact             @relation(fields: [contactId], references: [id])`,
`  contact        Contact             @relation(fields: [contactId], references: [id], onDelete: Restrict)
  journalEntry   JournalEntry?       @relation(fields: [journalEntryId], references: [id])`);

// 4. InvoiceLine relations
schema = schema.replace(
`  invoice     Invoice    @relation(fields: [invoiceId], references: [id], onDelete: Cascade)
  item        Item?      @relation(fields: [itemId], references: [id])
  warehouse   Warehouse? @relation(fields: [warehouseId], references: [id])`,
`  invoice     Invoice    @relation(fields: [invoiceId], references: [id], onDelete: Cascade)
  item        Item?      @relation(fields: [itemId], references: [id], onDelete: Restrict)
  warehouse   Warehouse? @relation(fields: [warehouseId], references: [id], onDelete: Restrict)
  account     ChartOfAccount @relation(fields: [accountId], references: [id])
  taxCode     TaxCode?       @relation(fields: [taxCodeId], references: [id])`);

// 5. Payment relations
schema = schema.replace(
`  allocations    PaymentAllocation[]`,
`  allocations    PaymentAllocation[]
  accountRef     ChartOfAccount      @relation("PaymentAccount", fields: [accountId], references: [id])
  whtAccountRef  ChartOfAccount?     @relation("PaymentWhtAccount", fields: [whtAccountId], references: [id])
  journalEntry   JournalEntry?       @relation(fields: [journalEntryId], references: [id])`);

// 6. FixedAsset relations
schema = schema.replace(
`  schedules             DepreciationSchedule[]`,
`  schedules             DepreciationSchedule[]
  assetAccount          ChartOfAccount       @relation("AssetAccount", fields: [assetAccountId], references: [id])
  depreciationAccount   ChartOfAccount       @relation("DepreciationAccount", fields: [depreciationAccountId], references: [id])
  expenseAccount        ChartOfAccount       @relation("ExpenseAccount", fields: [expenseAccountId], references: [id])`);

// 7. Item relations
schema = schema.replace(
`  InventoryTransaction InventoryTransaction[]`,
`  InventoryTransaction InventoryTransaction[]
  inventoryAccount     ChartOfAccount?        @relation("InventoryAccount", fields: [inventoryAccountId], references: [id])
  cogsAccount          ChartOfAccount?        @relation("CogsAccount", fields: [cogsAccountId], references: [id])
  revenueAccount       ChartOfAccount?        @relation("RevenueAccount", fields: [revenueAccountId], references: [id])`);

// 8. InventoryTransaction relations
schema = schema.replace(
`  warehouse      Warehouse                @relation(fields: [warehouseId], references: [id])`,
`  warehouse      Warehouse                @relation(fields: [warehouseId], references: [id])
  journalEntry   JournalEntry?            @relation(fields: [journalEntryId], references: [id])`);

// 9. PayrollRun relations
schema = schema.replace(
`  payslips              Payslip[]`,
`  payslips              Payslip[]
  journalEntry          JournalEntry? @relation(fields: [journalEntryId], references: [id])`);

// 10. BankStatementTransaction
schema = schema.replace(
`model BankStatementTransaction {
  id              String        @id @default(uuid())
  bankStatementId String`,
`model BankStatementTransaction {
  id              String        @id @default(uuid())
  tenantId        String
  bankStatementId String`);

schema = schema.replace(
`  bankStatement   BankStatement @relation(fields: [bankStatementId], references: [id], onDelete: Cascade)

  @@index([bankStatementId])
}`,
`  bankStatement   BankStatement @relation(fields: [bankStatementId], references: [id], onDelete: Cascade)
  tenant          Tenant        @relation(fields: [tenantId], references: [id])

  @@index([tenantId])
  @@index([bankStatementId])
}`);

// 11. DepreciationSchedule
schema = schema.replace(
`model DepreciationSchedule {
  id             String     @id @default(uuid())
  fixedAssetId   String`,
`model DepreciationSchedule {
  id             String     @id @default(uuid())
  tenantId       String
  fixedAssetId   String`);

schema = schema.replace(
`  fixedAsset     FixedAsset @relation(fields: [fixedAssetId], references: [id], onDelete: Cascade)

  @@index([fixedAssetId])
}`,
`  fixedAsset     FixedAsset @relation(fields: [fixedAssetId], references: [id], onDelete: Cascade)
  tenant         Tenant     @relation(fields: [tenantId], references: [id])

  @@index([tenantId])
  @@index([fixedAssetId])
}`);

// 12. InventoryBalance
schema = schema.replace(
`model InventoryBalance {
  id          String    @id @default(uuid())
  itemId      String`,
`model InventoryBalance {
  id          String    @id @default(uuid())
  tenantId    String
  itemId      String`);

schema = schema.replace(
`  warehouse   Warehouse @relation(fields: [warehouseId], references: [id])

  @@unique([itemId, warehouseId])
}`,
`  warehouse   Warehouse @relation(fields: [warehouseId], references: [id])
  tenant      Tenant    @relation(fields: [tenantId], references: [id])

  @@index([tenantId])
  @@unique([itemId, warehouseId])
}`);

// 13. InventoryLot
schema = schema.replace(
`model InventoryLot {
  id          String    @id @default(uuid())
  itemId      String`,
`model InventoryLot {
  id          String    @id @default(uuid())
  tenantId    String
  itemId      String`);

schema = schema.replace(
`  warehouse   Warehouse @relation(fields: [warehouseId], references: [id])

  @@index([itemId, warehouseId])
}`,
`  warehouse   Warehouse @relation(fields: [warehouseId], references: [id])
  tenant      Tenant    @relation(fields: [tenantId], references: [id])

  @@index([tenantId])
  @@index([itemId, warehouseId])
}`);

// 14. Payslip
schema = schema.replace(
`model Payslip {
  id                String     @id @default(uuid())
  payrollRunId      String`,
`model Payslip {
  id                String     @id @default(uuid())
  tenantId          String
  payrollRunId      String`);

schema = schema.replace(
`  payrollRun        PayrollRun @relation(fields: [payrollRunId], references: [id], onDelete: Cascade)

  @@unique([payrollRunId, employeeProfileId])
  @@index([payrollRunId])
}`,
`  payrollRun        PayrollRun @relation(fields: [payrollRunId], references: [id], onDelete: Cascade)
  tenant            Tenant     @relation(fields: [tenantId], references: [id])

  @@index([tenantId])
  @@unique([payrollRunId, employeeProfileId])
  @@index([payrollRunId])
}`);

// 15. PurchaseOrderLine
schema = schema.replace(
`model PurchaseOrderLine {
  id              String        @id @default(uuid())
  purchaseOrderId String`,
`model PurchaseOrderLine {
  id              String        @id @default(uuid())
  tenantId        String
  purchaseOrderId String`);

schema = schema.replace(
`  purchaseOrder   PurchaseOrder @relation(fields: [purchaseOrderId], references: [id], onDelete: Cascade)

  @@index([purchaseOrderId])
}`,
`  purchaseOrder   PurchaseOrder @relation(fields: [purchaseOrderId], references: [id], onDelete: Cascade)
  tenant          Tenant        @relation(fields: [tenantId], references: [id])

  @@index([tenantId])
  @@index([purchaseOrderId])
}`);

// 16. EmployeeProfile isSaudi
schema = schema.replace(
`  // KSA HR fields
  hireDate       DateTime?`,
`  // KSA HR fields
  isSaudi        Boolean   @default(true)
  hireDate       DateTime?`);

// Add required back-relations for JournalEntry
schema = schema.replace(
`  lines       JournalEntryLine[]`,
`  lines       JournalEntryLine[]
  invoice                 Invoice?
  payment                 Payment?
  inventoryTransaction    InventoryTransaction?
  payrollRun              PayrollRun?`);

fs.writeFileSync('apps/api/prisma/schema.prisma', schema, 'utf8');
console.log('Schema modified successfully.');

File: fix_schema.py
with open('apps/api/prisma/schema.prisma', 'r') as f:
    lines = f.readlines()

new_lines = []
current_model = None
for i, line in enumerate(lines):
    if line.startswith('model '):
        current_model = line.split()[1]
    
    # 1. RolePermission
    if current_model == 'RolePermission':
        if 'id           String     @id @default(uuid())' in line:
            new_lines.append(line)
            new_lines.append('  tenantId     String\n')
            continue
        if 'role         Role       @relation(fields: [roleId], references: [id])' in line:
            new_lines.append('  tenant       Tenant     @relation(fields: [tenantId], references: [id])\n')
            new_lines.append('  role         Role       @relation(fields: [roleId], references: [id], onDelete: Cascade)\n')
            continue
        if 'permission   Permission @relation(fields: [permissionId], references: [id])' in line:
            new_lines.append('  permission   Permission @relation(fields: [permissionId], references: [id], onDelete: Cascade)\n')
            continue
        if '@@unique([roleId, permissionId])' in line:
            new_lines.append(line)
            new_lines.append('  @@index([tenantId])\n')
            continue

    # 2. JournalEntryLine
    elif current_model == 'JournalEntryLine':
        if 'journalEntry     JournalEntry    @relation(fields: [journalEntryId], references: [id])' in line:
            new_lines.append('  journalEntry     JournalEntry    @relation(fields: [journalEntryId], references: [id], onDelete: Cascade)\n')
            continue
        if 'contact          Contact?        @relation(fields: [contactId], references: [id])' in line:
            new_lines.append('  contact          Contact?        @relation(fields: [contactId], references: [id], onDelete: Restrict)\n')
            continue

    # 3. Invoice
    elif current_model == 'Invoice':
        if 'contact        Contact             @relation(fields: [contactId], references: [id])' in line:
            new_lines.append('  contact        Contact             @relation(fields: [contactId], references: [id], onDelete: Restrict)\n')
            new_lines.append('  journalEntry   JournalEntry?       @relation(fields: [journalEntryId], references: [id])\n')
            continue

    # 4. InvoiceLine
    elif current_model == 'InvoiceLine':
        if 'item        Item?      @relation(fields: [itemId], references: [id])' in line:
            new_lines.append('  item        Item?      @relation(fields: [itemId], references: [id], onDelete: Restrict)\n')
            continue
        if 'warehouse   Warehouse? @relation(fields: [warehouseId], references: [id])' in line:
            new_lines.append('  warehouse   Warehouse? @relation(fields: [warehouseId], references: [id], onDelete: Restrict)\n')
            new_lines.append('  account     ChartOfAccount @relation(fields: [accountId], references: [id])\n')
            new_lines.append('  taxCode     TaxCode?       @relation(fields: [taxCodeId], references: [id])\n')
            continue

    # 5. Payment
    elif current_model == 'Payment':
        if 'contact        Contact             @relation(fields: [contactId], references: [id])' in line:
            new_lines.append('  contact        Contact             @relation(fields: [contactId], references: [id], onDelete: Restrict)\n')
            continue
        if 'allocations    PaymentAllocation[]' in line:
            new_lines.append(line)
            new_lines.append('  accountRef     ChartOfAccount      @relation("PaymentAccount", fields: [accountId], references: [id])\n')
            new_lines.append('  whtAccountRef  ChartOfAccount?     @relation("PaymentWhtAccount", fields: [whtAccountId], references: [id])\n')
            new_lines.append('  journalEntry   JournalEntry?       @relation(fields: [journalEntryId], references: [id])\n')
            continue

    # 6. FixedAsset
    elif current_model == 'FixedAsset':
        if 'schedules             DepreciationSchedule[]' in line:
            new_lines.append(line)
            new_lines.append('  assetAccount          ChartOfAccount       @relation("AssetAccount", fields: [assetAccountId], references: [id])\n')
            new_lines.append('  depreciationAccount   ChartOfAccount       @relation("DepreciationAccount", fields: [depreciationAccountId], references: [id])\n')
            new_lines.append('  expenseAccount        ChartOfAccount       @relation("ExpenseAccount", fields: [expenseAccountId], references: [id])\n')
            continue

    # 7. Item
    elif current_model == 'Item':
        if 'InventoryTransaction InventoryTransaction[]' in line:
            new_lines.append(line)
            new_lines.append('  inventoryAccount     ChartOfAccount?        @relation("InventoryAccount", fields: [inventoryAccountId], references: [id])\n')
            new_lines.append('  cogsAccount          ChartOfAccount?        @relation("CogsAccount", fields: [cogsAccountId], references: [id])\n')
            new_lines.append('  revenueAccount       ChartOfAccount?        @relation("RevenueAccount", fields: [revenueAccountId], references: [id])\n')
            continue

    # 8. InventoryTransaction
    elif current_model == 'InventoryTransaction':
        if 'warehouse      Warehouse                @relation(fields: [warehouseId], references: [id])' in line:
            new_lines.append(line)
            new_lines.append('  journalEntry   JournalEntry?            @relation(fields: [journalEntryId], references: [id])\n')
            continue

    # 9. PayrollRun
    elif current_model == 'PayrollRun':
        if 'payslips              Payslip[]' in line:
            new_lines.append(line)
            new_lines.append('  journalEntry          JournalEntry? @relation(fields: [journalEntryId], references: [id])\n')
            continue

    # 10. BankStatementTransaction
    elif current_model == 'BankStatementTransaction':
        if 'bankStatementId String' in line:
            new_lines.append('  tenantId        String\n')
            new_lines.append(line)
            continue
        if 'bankStatement   BankStatement @relation(fields: [bankStatementId], references: [id], onDelete: Cascade)' in line:
            new_lines.append(line)
            new_lines.append('  tenant          Tenant        @relation(fields: [tenantId], references: [id])\n')
            continue
        if '@@index([bankStatementId])' in line:
            new_lines.append('  @@index([tenantId])\n')
            new_lines.append(line)
            continue

    # 11. DepreciationSchedule
    elif current_model == 'DepreciationSchedule':
        if 'fixedAssetId   String' in line:
            new_lines.append('  tenantId       String\n')
            new_lines.append(line)
            continue
        if 'fixedAsset     FixedAsset @relation(fields: [fixedAssetId], references: [id], onDelete: Cascade)' in line:
            new_lines.append(line)
            new_lines.append('  tenant         Tenant     @relation(fields: [tenantId], references: [id])\n')
            continue
        if '@@index([fixedAssetId])' in line:
            new_lines.append('  @@index([tenantId])\n')
            new_lines.append(line)
            continue

    # 12. InventoryBalance
    elif current_model == 'InventoryBalance':
        if 'itemId      String' in line:
            new_lines.append('  tenantId    String\n')
            new_lines.append(line)
            continue
        if 'warehouse   Warehouse @relation(fields: [warehouseId], references: [id])' in line:
            new_lines.append(line)
            new_lines.append('  tenant      Tenant    @relation(fields: [tenantId], references: [id])\n')
            continue
        if '@@unique([itemId, warehouseId])' in line:
            new_lines.append('  @@index([tenantId])\n')
            new_lines.append(line)
            continue

    # 13. InventoryLot
    elif current_model == 'InventoryLot':
        if 'itemId      String' in line:
            new_lines.append('  tenantId    String\n')
            new_lines.append(line)
            continue
        if 'warehouse   Warehouse @relation(fields: [warehouseId], references: [id])' in line:
            new_lines.append(line)
            new_lines.append('  tenant      Tenant    @relation(fields: [tenantId], references: [id])\n')
            continue
        if '@@index([itemId, warehouseId])' in line:
            new_lines.append('  @@index([tenantId])\n')
            new_lines.append(line)
            continue

    # 14. Payslip
    elif current_model == 'Payslip':
        if 'payrollRunId      String' in line:
            new_lines.append('  tenantId          String\n')
            new_lines.append(line)
            continue
        if 'payrollRun        PayrollRun @relation(fields: [payrollRunId], references: [id], onDelete: Cascade)' in line:
            new_lines.append(line)
            new_lines.append('  tenant            Tenant     @relation(fields: [tenantId], references: [id])\n')
            continue
        if '@@unique([payrollRunId, employeeProfileId])' in line:
            new_lines.append('  @@index([tenantId])\n')
            new_lines.append(line)
            continue

    # 15. PurchaseOrderLine
    elif current_model == 'PurchaseOrderLine':
        if 'purchaseOrderId String' in line:
            new_lines.append('  tenantId        String\n')
            new_lines.append(line)
            continue
        if 'purchaseOrder   PurchaseOrder @relation(fields: [purchaseOrderId], references: [id], onDelete: Cascade)' in line:
            new_lines.append(line)
            new_lines.append('  tenant          Tenant        @relation(fields: [tenantId], references: [id])\n')
            continue
        if '@@index([purchaseOrderId])' in line:
            new_lines.append('  @@index([tenantId])\n')
            new_lines.append(line)
            continue

    # 16. EmployeeProfile
    elif current_model == 'EmployeeProfile':
        if 'hireDate       DateTime?' in line:
            new_lines.append('  isSaudi        Boolean   @default(true)\n')
            new_lines.append(line)
            continue

    # 17. JournalEntry
    elif current_model == 'JournalEntry':
        if 'lines       JournalEntryLine[]' in line:
            new_lines.append(line)
            new_lines.append('  invoice                 Invoice?\n')
            new_lines.append('  payment                 Payment?\n')
            new_lines.append('  inventoryTransaction    InventoryTransaction?\n')
            new_lines.append('  payrollRun              PayrollRun?\n')
            continue

    new_lines.append(line)

with open('apps/api/prisma/schema.prisma', 'w') as f:
    f.writelines(new_lines)
